oauth2client/flask.py

```diff
@@ -0,0 +1,329 @@
+        if client_type != clientsecrets.TYPE_WEB:
+            raise ValueError(
+                "The flow specified in %s isn't supported." % client_type)
+
+        return client_info['client_id'], client_info['client_secret']
+
+    def _load_app_config(self, app):
+        if ('OAUTH2_CLIENT_SECRET' not in app.config or
+                'OAUTH2_CLIENT_ID' not in app.config):
+            raise ValueError(
+                "OAuth2 configuration could not be found. Either "
+                "OAUTH2_CLIENT_SECRETS_JSON or OAUTH2_CLIENT_ID and "
+                "OAUTH2_CLIENT_SECRET must be specified.")
+
+        return (app.config['OAUTH2_CLIENT_ID'],
+                app.config['OAUTH2_CLIENT_SECRET'])
+
+    def _update_session(self, resp):
+        """Ensures that the session's copy of the credentials is updated
+        when the credential refreshes its access token."""
+        ctx = _app_ctx_stack.top
+        if hasattr(ctx, 'oauth2_credentials'):
+            session['oauth2credentials'] = ctx.oauth2_credentials.to_json()
+        return resp
+
+    def _request_user_info(self):
+        """Gets the user's profile information from Google+
+
+        Makes an HTTP request to the Google+ API to retrieve the user's basic
+        profile information, including full name and photo.
+        """
+        http = self.http()
+        resp, content = http.request(
+            'https://www.googleapis.com/plus/v1/people/me')
+
+        if resp.status != 200:
+            current_app.logger.error(
+                "Error while obtaining user profile: %s" % resp)
+            return None
+
+        try:
+            profile = json.loads(content)
+        except ValueError:
+            current_app.logger.error("Invalid user profile json: %s", content)
+            return None
+
+        return profile
+
+    def _make_flow(self):
+        # Generate a CSRF token to prevent malicious requests.
+        state = ''.join(random.choice(string.ascii_uppercase + string.digits)
+                        for _ in range(32))
+
+        session['oauth2state'] = state
+
+        return OAuth2WebServerFlow(
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            scope=self.scopes,
+            state=state,
+            approval_prompt='force',  # TODO: validate intended behavior
+            redirect_uri=url_for('oauth2.callback', _external=True))
+
+    def _create_blueprint(self):
+        bp = Blueprint('oauth2', __name__)
+        bp.add_url_rule('/oauth2authorize', 'authorize', self.authorize_view)
+        bp.add_url_rule('/oauth2logout', 'logout', self.logout_view)
+        bp.add_url_rule('/oauth2callback', 'callback', self.callback_view)
+
+        return bp
+
+    def authorize_view(self):
+        return_url = request.args.get('return_url')
+        if not return_url:
+            return_url = request.referrer or '/'
+        session['oauth2return'] = return_url
```

Comment:
So, this seems OK, but slightly less robust than using the callback state. Where this might break down is if a user had multiple tabs open for the same website, and the consent to a prompt from one page might go to another. Not the end of the world, but it may be worth seeing if this can be put into state using a dictionary of json or something instead.